janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java

```diff
@@ -509,42 +509,57 @@ private void commitDeletes(String collectionName, List<String> deleteIds) throws
-        List<RawQuery.Result<String>> result;
-        String collection = query.getStore();
-        String keyIdField = getKeyFieldId(collection);
+    private QueryResponse runCommonQuery(RawQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx, String collection, String keyIdField) throws BackendException {
         SolrQuery solrQuery = new SolrQuery(query.getQuery())
                                 .addField(keyIdField)
                                 .setIncludeScore(true)
                                 .setStart(query.getOffset())
                                 .setRows(query.hasLimit() ? query.getLimit() : maxResults);
-
         try {
-            QueryResponse response = solrClient.query(collection, solrQuery);
-            if (logger.isDebugEnabled())
-                logger.debug("Executed query [{}] in {} ms", query.getQuery(), response.getElapsedTime());
-
-            int totalHits = response.getResults().size();
-            if (!query.hasLimit() && totalHits >= maxResults) {
-                logger.warn("Query result set truncated to first [{}] elements for query: {}", maxResults, query);
-            }
-            result = new ArrayList<RawQuery.Result<String>>(totalHits);
-
-            for (SolrDocument hit : response.getResults()) {
-                double score = Double.parseDouble(hit.getFieldValue("score").toString());
-                result.add(new RawQuery.Result<String>(hit.getFieldValue(keyIdField).toString(), score));
-            }
+        	QueryResponse response = solrClient.query(collection, solrQuery);
+        	return response;
         } catch (IOException e) {
             logger.error("Query did not complete : ", e);
             throw new PermanentBackendException(e);
         } catch (SolrServerException e) {
             logger.error("Unable to query Solr index.", e);
             throw new PermanentBackendException(e);
         }
+    }
+    
+    @Override
+    public Iterable<RawQuery.Result<String>> query(RawQuery query, KeyInformation.IndexRetriever informations, BaseTransaction tx) throws BackendException {
+    	String collection = query.getStore();
+    	String keyIdField = getKeyFieldId(collection);
+    	List<RawQuery.Result<String>> result;
+    	QueryResponse response = runCommonQuery(query, informations, tx, collection, keyIdField);
+        if (logger.isDebugEnabled())
+            logger.debug("Executed query [{}] in {} ms", query.getQuery(), response.getElapsedTime());
+
+        int totalHits = response.getResults().size();
+        if (!query.hasLimit() && totalHits >= maxResults) {
+            logger.warn("Query result set truncated to first [{}] elements for query: {}", maxResults, query);
+        }
+        result = new ArrayList<RawQuery.Result<String>>(totalHits);
+        for (SolrDocument hit : response.getResults()) {
```

Comment:
response.getResults().stream().map(to a RawQuery.Result).collect(Collectors.toList())